src/ast/passes/resource_analyser.cpp

```diff
@@ -303,6 +333,33 @@ void ResourceAnalyser::visit(For &f)
   }
 }
 
+void ResourceAnalyser::visit(AssignMapStatement &assignment)
+{
+  // Unfortunately, CodegenLLVM doesn't recurse into assignment.map
```

Comment:
I'll be honest. I read this comment twice and I'm still not sure what's happening here - maybe I just need a brain restart.

Reply:
1. CodegenLLVM module traverses AST via accept() method
2. Visitor module traverse AST via visit() method
3. CodegenLLVM will traverse AssignmentMap -> accept(assignment.expr) and accept(assignment.map.key_expr) BUT will not call accept(assignment.map)
4. However, Visitor will traverse AssignmentMap -> visit(assignment.expr) and visit(assignment.map) 
5. CodegenLLVM module will do a map read lookup when traversing AST into assignment.map

Thus, to avoid over-counting read buffer, we need to do this hack above. Unfortunately, this optimization is pretty important, otherwise a bunch of codegen tests start allocating buffers needlessly. So I'll try to figure out how to clarify it.

Reply:
@jordalgo updated the comment with diagrams, let me know if its not clear still or you have suggestions